pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/SegmentSizeRule.java

```diff
@@ -0,0 +1,145 @@
+      File generatedSegmentDir =
+          new MemoryEstimator.SegmentGenerator(_input._schemaWithMetaData, _input._schema, tableConfig,
+              numRowsInGeneratedSegment, true).generate();
+      segmentSize = FileUtils.sizeOfDirectory(generatedSegmentDir);
+      numRows = numRowsInGeneratedSegment;
+
+      // cleanup
+      try {
+        FileUtils.deleteDirectory(generatedSegmentDir);
+      } catch (IOException e) {
+        throw new RuntimeException("Cannot delete the generated segment directory", e);
+      }
+    } else {
+      segmentSize = segmentSizeRuleParams.getActualSegmentSizeMB() * MEGA_BYTE;
+      numRows = segmentSizeRuleParams.getNumRowsInActualSegment();
+    }
+
+    // estimate optimal segment count & size parameters
+    SegmentSizeRecommendations params =
+        estimate(segmentSize, segmentSizeRuleParams.getDesiredSegmentSizeMB() * MEGA_BYTE, numRows,
+            _input.getNumRecordsPerPush());
+
+    // wire the recommendations and also update the cardinalities in input manager. The updated cardinalities are used in subsequent rules.
+    _output.setSegmentSizeRecommendations(params);
+    _input.capCardinalities((int) params.getNumRowsPerSegment());
+  }
+
+  /**
+   * Estimate segment size parameters by extrapolation based on the number of records and size of the generated segment.
+   * The linear extrapolation used here is not optimal because of columnar way of storing data and usage of different
+   * indices. Another way would be to iteratively generate new segments with expected number of rows until the ideal
+   * segment is found, but that's costly because of the time it takes to generate segments. Although the extrapolation
+   * approach seems to be less accurate, it is chosen due to its performance.
+   *
+   * @param GSS  generated segment size
+   * @param DSS  desired segment size
+   * @param NRGS num records of generated segment
+   * @param NRPP num records per push
+   * @return recommendations on optimal segment count, size, and number of records
+   */
+  @VisibleForTesting
+  SegmentSizeRecommendations estimate(long GSS, int DSS, int NRGS, long NRPP) {
+
+    // calc num rows in desired segment
+    double sizeRatio = (double) DSS / GSS;
+    long numRowsInDesiredSegment = Math.round(NRGS * sizeRatio);
+
+    // calc optimal num segment
+    long optimalNumSegments = Math.round(NRPP / (double) numRowsInDesiredSegment);
+    optimalNumSegments = Math.max(optimalNumSegments, 1);
+
+    // revise optimal num rows in segment
+    long optimalNumRowsInSegment = Math.round(NRPP / (double) optimalNumSegments);
+
+    // calc optimal segment size
+    double rowRatio = (double) optimalNumRowsInSegment / NRGS;
+    long optimalSegmentSize = Math.round(GSS * rowRatio);
+
+    return new SegmentSizeRecommendations(optimalNumRowsInSegment, optimalNumSegments, optimalSegmentSize);
+  }
+
+  private TableConfig createTableConfig(Schema schema) {
```

Comment:
We should mention in the javadoc that this rule is run before other rules, we run into chicken-egg problem. Index recommendation, dictionary, bloom filter rules will be run next (once this rule computes CPS). Only after the subsequent rules run and the index recommendation is available, will we know what segment will look like. So here we just assume dictionary on all dimensions and raw for metric columns and try to generate a segment quickly. This also means that value computed at line 134 should ideally be bumped up by (say 20%) to account for what the size might look like when all the index/dictionary/bloom recommendations are applied.

Reply:
Agree 👍 
Updated the doc.